File: tests/cross_driver/node/types_smoke.js
// Cross-driver BSON type fidelity smoke — Node (mongo-node-driver).
//
// Same workload as tests/cross_driver/go/types/main.go: insert one
// document containing every BSON type whose JS-side representation is
// a distinct strict class, then find it back and assert each field
// preserved its class and value. node-driver's BSON layer
// (`bson` npm package) is a separate implementation from pymongo's,
// so any wire-shape divergence between SecantusDB and a real mongod
// trips here with a class-mismatch.

const { MongoClient, ObjectId, Decimal128, Long, Binary, Int32, Double } = require('mongodb');

function fail(msg) {
  console.error(msg);
  process.exit(1);
}

async function main() {
  const uri = process.env.MONGODB_URI;
  if (!uri) fail('MONGODB_URI not set');

  const client = new MongoClient(uri, { serverSelectionTimeoutMS: 5000 });
  await client.connect();
  try {
    const coll = client.db('types_xd').collection('c');
    await coll.drop().catch(() => {});

    const objID = new ObjectId();
    const dec = Decimal128.fromString('3.141592653589793238');
    const when = new Date('2026-05-06T12:34:56.789Z');
    const bin = new Binary(Buffer.from('hello'), 0);

    // Force int32 vs int64 explicitly — JS's Number is 64-bit float, so
    // SecantusDB must respect the BSON-level int32 / int64 tag we send,
    // not silently re-tag it from value alone.
    const doc = {
      _id: objID,
      i32: new Int32(2147483647),
      i64: Long.fromString('9223372036854775807'),
      f64: new Double(2.5),
      dec,
      dt: when,
      bin,
      b: true,
      n: null,
      sub: { x: new Int32(1) },
      arr: [new Int32(1), 'two', new Double(3.5)],
    };

    await coll.insertOne(doc);

    // Use { promoteValues: false } at the collection level so int32 /
    // int64 / Double come back as their wrapper classes rather than
    // collapsed to JS Number — the whole point of the test.
    const collTyped = client
      .db('types_xd')
      .collection('c', { promoteValues: false, promoteLongs: false });
    const got = await collTyped.findOne({ _id: objID });
    if (!got) fail('findOne returned null');

    if (!(got._id instanceof ObjectId) || !got._id.equals(objID)) {
      fail(`_id: got ${got._id} (${got._id?.constructor?.name})`);
    }
    if (!(got.i32 instanceof Int32) || got.i32.valueOf() !== 2147483647) {
      fail(`i32: got ${got.i32} (${got.i32?.constructor?.name})`);
    }
    if (!(got.i64 instanceof Long) || got.i64.toString() !== '9223372036854775807') {
      fail(`i64: got ${got.i64} (${got.i64?.constructor?.name})`);
    }
    if (!(got.f64 instanceof Double) || got.f64.valueOf() !== 2.5) {
      fail(`f64: got ${got.f64} (${got.f64?.constructor?.name})`);
    }
    if (!(got.dec instanceof Decimal128) || got.dec.toString() !== dec.toString()) {
      fail(`dec: got ${got.dec} (${got.dec?.constructor?.name})`);
    }
    if (!(got.dt instanceof Date) || got.dt.getTime() !== when.getTime()) {
      fail(`dt: got ${got.dt}`);
    }
    if (
      !(got.bin instanceof Binary) ||
      got.bin.sub_type !== 0 ||
      Buffer.from(got.bin.value()).toString('utf8') !== 'hello'
    ) {
      fail(`bin: got ${JSON.stringify(got.bin)}`);
    }
    if (got.b !== true) fail(`b: got ${got.b}`);
    if (got.n !== null) fail(`n: got ${got.n}`);
    if (!(got.sub.x instanceof Int32) || got.sub.x.valueOf() !== 1) {
      fail(`sub.x: got ${got.sub.x} (${got.sub.x?.constructor?.name})`);
    }
    if (!Array.isArray(got.arr) || got.arr.length !== 3) {
      fail(`arr: got ${JSON.stringify(got.arr)}`);
    }
    if (!(got.arr[0] instanceof Int32) || got.arr[0].valueOf() !== 1) {
      fail(`arr[0]: got ${got.arr[0]} (${got.arr[0]?.constructor?.name})`);
    }
    if (got.arr[1] !== 'two') fail(`arr[1]: got ${got.arr[1]}`);
    if (!(got.arr[2] instanceof Double) || got.arr[2].valueOf() !== 3.5) {
      fail(`arr[2]: got ${got.arr[2]} (${got.arr[2]?.constructor?.name})`);
    }

    console.log('OK');
  } finally {
    await client.close();
  }
}

main().catch((err) => fail(`uncaught: ${err && err.stack ? err.stack : err}`));

File: tests/cross_driver/rust/src/bin/types_smoke.rs
//! Cross-driver BSON type fidelity smoke — Rust (mongo-rust-driver).
//!
//! Insert one document containing every BSON type the Rust driver
//! surfaces as a distinct variant of `bson::Bson`, then find it back
//! and assert each field preserved its variant and value. Catches
//! wire-shape divergences that one driver tolerates but another
//! doesn't (e.g. int32-vs-int64 collision under permissive type
//! coercion).
use std::env;
use std::process::exit;

use bson::oid::ObjectId;
use bson::spec::BinarySubtype;
use bson::{doc, Binary, Bson, DateTime, Decimal128};
use mongodb::sync::Client;

fn fail(msg: String) -> ! {
    eprintln!("{msg}");
    exit(1);
}

fn main() {
    let uri = env::var("MONGODB_URI").unwrap_or_else(|_| {
        eprintln!("MONGODB_URI not set");
        exit(2);
    });

    let client = Client::with_uri_str(&uri).unwrap_or_else(|e| fail(format!("connect: {e}")));
    let coll = client.database("types_xd_rust").collection::<bson::Document>("c");
    let _ = coll.drop().run();

    let oid = ObjectId::new();
    let dec = "3.141592653589793238".parse::<Decimal128>().unwrap();
    let when = DateTime::from_millis(1_780_000_000_000); // 2026-05-29
    let bin_payload = b"hello".to_vec();
    let bin = Binary { subtype: BinarySubtype::Generic, bytes: bin_payload.clone() };

    let doc_in = doc! {
        "_id": oid,
        "i32": 2_147_483_647_i32,
        "i64": 9_223_372_036_854_775_807_i64,
        "f64": 2.5_f64,
        "dec": dec.clone(),
        "dt": when,
        "bin": bin.clone(),
        "b": true,
        "n": Bson::Null,
        "sub": doc! { "x": 1_i32 },
        "arr": [Bson::Int32(1), Bson::String("two".into()), Bson::Double(3.5)],
    };
    coll.insert_one(doc_in).run().unwrap_or_else(|e| fail(format!("insert: {e}")));

    let got = coll
        .find_one(doc! {"_id": oid})
        .run()
        .unwrap_or_else(|e| fail(format!("find: {e}")))
        .unwrap_or_else(|| fail("find_one returned None".into()));

    if got.get_object_id("_id").ok() != Some(oid) {
        fail(format!("_id: got {:?}", got.get("_id")));
    }
    if got.get_i32("i32").ok() != Some(2_147_483_647) {
        fail(format!("i32: got {:?}", got.get("i32")));
    }
    if got.get_i64("i64").ok() != Some(9_223_372_036_854_775_807) {
        fail(format!("i64: got {:?}", got.get("i64")));
    }
    if got.get_f64("f64").ok() != Some(2.5) {
        fail(format!("f64: got {:?}", got.get("f64")));
    }
    if got.get("dec") != Some(&Bson::Decimal128(dec)) {
        fail(format!("dec: got {:?}", got.get("dec")));
    }
    if got.get_datetime("dt").ok().map(|d| d.timestamp_millis()) != Some(1_780_000_000_000) {
        fail(format!("dt: got {:?}", got.get("dt")));
    }
    match got.get("bin") {
        Some(Bson::Binary(b)) if b.bytes == bin_payload && b.subtype == BinarySubtype::Generic => {}
        other => fail(format!("bin: got {:?}", other)),
    }
    if got.get_bool("b").ok() != Some(true) {
        fail(format!("b: got {:?}", got.get("b")));
    }
    if got.get("n") != Some(&Bson::Null) {
        fail(format!("n: got {:?}", got.get("n")));
    }
    let sub = got.get_document("sub").unwrap_or_else(|e| fail(format!("sub: {e}")));
    if sub.get_i32("x").ok() != Some(1) {
        fail(format!("sub.x: got {:?}", sub.get("x")));
    }
    let arr = got.get_array("arr").unwrap_or_else(|e| fail(format!("arr: {e}")));
    if arr.len() != 3 {
        fail(format!("arr len: got {}", arr.len()));
    }
    match (&arr[0], &arr[1], &arr[2]) {
        (Bson::Int32(1), Bson::String(s), Bson::Double(3.5)) if s == "two" => {}
        _ => fail(format!("arr: got {:?}", arr)),
    }

    println!("OK");
}
